aten/src/ATen/native/cuda/UpSampleBilinear2d.cu

```diff
@@ -1,25 +1,312 @@
+// Adapted from interp.cpp from Caffe util by Pauline Luc
+// Originally developed by George Papandreou
 #include <ATen/ATen.h>
+#include <ATen/AccumulateType.h>
 #include <ATen/NativeFunctions.h>
-#include <ATen/LegacyTHFunctions.h>
+#include <ATen/TensorUtils.h>
+#include <ATen/Utils.h>
+#include <ATen/cuda/CUDAContext.h>
+#include <ATen/cuda/CUDAApplyUtils.cuh>
+#include <ATen/native/cuda/UpSample.cuh>
 
 namespace at {
 namespace native {
+namespace {
+
+template <typename scalar_t, typename accscalar_t>
+#ifdef __HIP_PLATFORM_HCC__
+C10_LAUNCH_BOUNDS_1(1024)
+#endif
+__global__ void upsample_bilinear2d_out_frame(
+    const int64_t n,
+    const accscalar_t rheight,
+    const accscalar_t rwidth,
+    const bool align_corners,
+    const PackedTensorAccessor<scalar_t, 4> idata,
+    PackedTensorAccessor<scalar_t, 4> odata) {
+  int64_t index = threadIdx.x + blockIdx.x * blockDim.x;
+
+  const int64_t batchsize = idata.size(0);
+  const int64_t channels = idata.size(1);
+  const int64_t height1 = idata.size(2);
+  const int64_t width1 = idata.size(3);
+  const int64_t height2 = odata.size(2);
+  const int64_t width2 = odata.size(3);
+
+  if (index < n) {
+    const int64_t w2 = index % width2; // 0:width2-1
+    const int64_t h2 = index / width2; // 0:height2-1
+    // special case: just copy
+    if (height1 == height2 && width1 == width2) {
+      const int64_t h1 = h2;
+      const int64_t w1 = w2;
+      for (int n = 0; n < batchsize; n++) {
+        for (int c = 0; c < channels; ++c) {
+          const scalar_t val = idata[n][c][h1][w1];
+          odata[n][c][h2][w2] = val;
+        }
+      }
+      return;
+    }
+    //
+    const accscalar_t h1r = area_pixel_compute_source_index<accscalar_t>(
+        rheight, h2, align_corners, /*cubic=*/false);
+    const int64_t h1 = h1r;
```

Comment:
Use of `int64_t` here is much less benign from a performance perspective, as you're going to index with it.